File: python/secret-scan.py
#!/usr/bin/env python3
# secret-scan.mjs — portiert nach python
# Quelle: javascript, Onboarding@main:scripts/secret-scan.mjs
# Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

import asyncio
import os
import re
from pathlib import Path


root = Path(__file__).parent.parent
skipped = {
    "node_modules",
    ".next",
    ".git",
    ".pytest_cache",
    "__pycache__",
    "media-production/raw",
    "media-production/private",
}
patterns = [
    re.compile(r"sk-(?:proj|svcacct|ant|or-v1|admin)-[A-Za-z0-9_-]{20,}"),
    re.compile(r"(?:nvapi|lin_api|ntn|vcp)_[A-Za-z0-9_-]{20,}"),
    re.compile(r"ELEVENLABS_API_KEY\s*=\s*[\"']?[A-Za-z0-9]{20,}"),
    re.compile(r"WAVESPEED_API_KEY\s*=\s*[\"']?[A-Za-z0-9]{20,}"),
]
findings = []


async def walk(directory, relative=""):
    for entry in os.listdir(directory):
        rel = os.path.join(relative, entry)
        if (
            entry == ".env"
            or (entry.startswith(".env.") and entry != ".env.example")
            or any(
                rel == item
                or rel.startswith(f"{item}{os.sep}")
                or item in rel.split(os.sep)
                for item in skipped
            )
        ):
            continue
        target = os.path.join(directory, entry)
        if os.path.isdir(target):
            await walk(target, rel)
        elif os.path.getsize(target) < 2_000_000:
            try:
                with open(target, "r", encoding="utf-8") as f:
                    content = f.read()
            except Exception:
                content = ""
            for pattern in patterns:
                if pattern.search(content):
                    findings.append(rel)


async def main():
    await walk(root)
    unique_findings = list(set(findings))
    if unique_findings:
        print(f"Secret-Scan fehlgeschlagen: {', '.join(unique_findings)}", file=sys.stderr)
        sys.exit(1)
    print("Secret-Scan bestanden.")


if __name__ == "__main__":
    import sys
    asyncio.run(main())

File: perl5/secret-scan.pl
#!/usr/bin/perl
# secret-scan.mjs — portiert nach perl5
# Quelle: javascript, Onboarding@main:scripts/secret-scan.mjs
# Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use File::Find;
use File::Spec;
use Cwd qw(abs_path);

# Wurzelverzeichnis bestimmen
my $root = abs_path(File::Spec->catdir(__FILE__, '..', '..'));

# Verzeichnisse und Dateien, die übersprungen werden sollen
my %skipped = map { $_ => 1 } (
    "node_modules",
    ".next",
    ".git",
    ".pytest_cache",
    "__pycache__",
    "media-production/raw",
    "media-production/private"
);

# Muster für geheime Schlüssel
my @patterns = (
    qr/sk-(?:proj|svcacct|ant|or-v1|admin)-[A-Za-z0-9_-]{20,}/,
    qr/(?:nvapi|lin_api|ntn|vcp)_[A-Za-z0-9_-]{20,}/,
    qr/ELEVENLABS_API_KEY\s*=\s*["']?[A-Za-z0-9]{20,}/,
    qr/WAVESPEED_API_KEY\s*=\s*["']?[A-Za-z0-9]{20,}/,
);

# Funde sammeln
my %findings;

# Callback-Funktion für File::Find
sub scan_file {
    my $file = $_;
    my $full_path = $File::Find::name;
    my $relative_path = File::Spec->abs2rel($full_path, $root);

    # Prüfen, ob die Datei übersprungen werden soll
    foreach my $skip (keys %skipped) {
        if ($relative_path eq $skip || $relative_path =~ /^$skip\//) {
            $File::Find::prune = 1; # Verzeichnis überspringen
            return;
        }
    }

    # .env-Dateien prüfen
    if ($file eq '.env' || ($file =~ /^\.env\./ && $file ne '.env.example')) {
        $findings{$relative_path} = 1;
        return;
    }

    # Dateigröße prüfen (< 2 MB)
    if (-f $full_path && -s $full_path < 2_000_000) {
        open(my $fh, '<', $full_path) or return;
        my $content = do { local $/; <$fh> };
        close($fh);

        # Auf Muster prüfen
        foreach my $pattern (@patterns) {
            if ($content =~ /$pattern/) {
                $findings{$relative_path} = 1;
                last;
            }
        }
    }
}

# Verzeichnis durchsuchen
find(\&scan_file, $root);

# Ergebnisse auswerten
if (%findings) {
    my @unique_findings = keys %findings;
    print STDERR "Secret-Scan fehlgeschlagen: " . join(", ", @unique_findings) . "\n";
    exit(1);
}

print "Secret-Scan bestanden.\n";
